### PR TITLE
Multiple checks, fix for url encoding, TCP/Command checks

### PR DESCRIPTION
**Pull request by wendigo**
Addresses #61, #67

### CODE REVIEW

File: consul/consul.go [hunk 1]
```diff
@@ -203,26 +205,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+			})
+		default:
+			log.WithField("Id", task.AppID.String()).WithField("Address", serviceAddress).
+				Error(fmt.Printf("Unrecognized check protocol %s", check.Protocol))
 		}
+
 	}
-	return nil
+
+	spew.Dump(checks)
```
Comment (janisz) [inline consul/consul.go]: Do we need this?

File: consul/consul.go [hunk 2]
```diff
@@ -203,26 +204,47 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
-		Check:   c.marathonToConsulCheck(task, app.HealthChecks, serviceAddress),
+		Checks:  c.marathonToConsulChecks(task, app.HealthChecks, serviceAddress),
 	}, nil
 }
 
-func (c *Consul) marathonToConsulCheck(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) *consulapi.AgentServiceCheck {
-	//	TODO: Handle all types of checks
+func (c *Consul) marathonToConsulChecks(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) consulapi.AgentServiceChecks {
+
+	var checks consulapi.AgentServiceChecks = make(consulapi.AgentServiceChecks, 0)
```
Comment (janisz) [inline consul/consul.go]: We now required capacity at this point

File: consul/consul.go (hunk 1 above)
Comment (wendigo) [inline consul/consul.go]: Hold on :) I'm trying to test this PR as on my macbook tests are always failing :palm_tree:

File: consul/consul.go (hunk 2 above)
Comment (wendigo) [inline consul/consul.go]: Sure thing :) Fixing

**Comment by wendigo** [comment]
Ready to be review :pray:

File: consul/consul_test.go [hunk 3]
```diff
@@ -507,60 +507,43 @@ func TestDeregisterServices_shouldReturnErrorOnFailure(t *testing.T) {
...
+				IntervalSeconds:        40,
+				TimeoutSeconds:         20,
+				MaxConsecutiveFailures: 3,
+			},
+			{
+				Protocol: "COMMAND",
+				Command: struct {
+					Value string `json:"value`
+				}{Value: "echo 1"},
+				IntervalSeconds:        30,
```
Comment (janisz) [inline consul/consul_test.go]: Maybe add one unknown check to show it'll be ignored

Comment (wendigo) [inline consul/consul_test.go]: OK :)

File: consul/consul.go [hunk 4]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/allegro/marathon-consul/metrics"
 	"github.com/allegro/marathon-consul/utils"
 	consulapi "github.com/hashicorp/consul/api"
+	"strings"
```
Comment (janisz) [inline consul/consul.go]: Run goimports on this file or just move this line

Comment (wendigo) [inline consul/consul.go]: done

File: consul/consul_test.go [hunk 5]
```diff
@@ -585,10 +576,18 @@ func TestMarathonTaskToConsulServiceMapping(t *testing.T) {
...
+
+	assert.Equal(t, "http://127.0.0.6:8090/api/health?with=query", service.Checks[0].HTTP)
+	assert.Equal(t, "https://127.0.0.6:8090/secure/health?with=query", service.Checks[1].HTTP)
+	assert.Equal(t, "127.0.0.6:8443", service.Checks[2].TCP)
+	assert.Equal(t, "echo 1", service.Checks[3].Script)
+
+	assert.Equal(t, "60s", service.Checks[0].Interval)
+	assert.Equal(t, "50s", service.Checks[1].Interval)
+	assert.Equal(t, "40s", service.Checks[2].Interval)
+	assert.Equal(t, "30s", service.Checks[3].Interval)
```
Comment (janisz) [inline consul/consul_test.go]: How about comparing whole slices instead of specific fileds? I mean one assert for all checks?

Comment (wendigo) [inline consul/consul_test.go]: Done :)

File: consul/consul_test.go (hunk 3 above)
Comment (janisz) [inline consul/consul_test.go]: Missing one test case with HTTP protocol and invalid path

File: consul/consul.go [hunk 6]
```diff
@@ -203,26 +205,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
 		Tags:    c.marathonLabelsToConsulTags(app.Labels),
-		Check:   c.marathonToConsulCheck(task, app.HealthChecks, serviceAddress),
+		Checks:  c.marathonToConsulChecks(task, app.HealthChecks, serviceAddress),
 	}, nil
 }
 
-func (c *Consul) marathonToConsulCheck(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) *consulapi.AgentServiceCheck {
-	//	TODO: Handle all types of checks
+func (c *Consul) marathonToConsulChecks(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) consulapi.AgentServiceChecks {
+
```
Comment (janisz) [inline consul/consul.go]: I found newlines in this file inconsistent.

File: consul/consul.go [hunk 7]
```diff
@@ -203,26 +205,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+			if parsedUrl, err := url.Parse(check.Path); err == nil {
+				parsedUrl.Scheme = strings.ToLower(check.Protocol)
+				parsedUrl.Host = fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex])
+
+				checks = append(checks, &consulapi.AgentServiceCheck{
+					HTTP:     parsedUrl.String(),
+					Interval: fmt.Sprintf("%ds", check.IntervalSeconds),
+					Timeout:  fmt.Sprintf("%ds", check.TimeoutSeconds),
+				})
```
Comment (janisz) [inline consul/consul.go]: Log warning message like in default

File: consul/consul.go [hunk 8]
```diff
@@ -203,26 +205,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+
+		case "COMMAND":
+			checks = append(checks, &consulapi.AgentServiceCheck{
+				Script:   check.Command.Value,
+				Interval: fmt.Sprintf("%ds", check.IntervalSeconds),
+				Timeout:  fmt.Sprintf("%ds", check.TimeoutSeconds),
+			})
+		default:
+			log.WithField("Id", task.AppID.String()).WithField("Address", serviceAddress).
+				Error(fmt.Printf("Unrecognized check protocol %s", check.Protocol))
```
Comment (janisz) [inline consul/consul.go]: We usually use error to indicate request is not processed further. Please log warning here

File: consul/consul.go [hunk 9]
```diff
@@ -203,26 +205,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+			if parsedUrl, err := url.Parse(check.Path); err == nil {
+				parsedUrl.Scheme = strings.ToLower(check.Protocol)
+				parsedUrl.Host = fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex])
+
+				checks = append(checks, &consulapi.AgentServiceCheck{
+					HTTP:     parsedUrl.String(),
+					Interval: fmt.Sprintf("%ds", check.IntervalSeconds),
+					Timeout:  fmt.Sprintf("%ds", check.TimeoutSeconds),
+				})
+			}
```
Comment (janisz) [inline consul/consul.go]: Log warning message if error exists like below

File: consul/consul.go (hunk 8 above)
Comment (wendigo) [inline consul/consul.go]: OK. Done.

File: consul/consul.go (hunk 9 above)
Comment (wendigo) [inline consul/consul.go]: OK. Done.

File: consul/consul.go (hunk 6 above)
Comment (wendigo) [inline consul/consul.go]: Removed.

File: consul/consul.go [hunk 10]
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"net/url"
 
+	"strings"
```
Comment (janisz) [inline consul/consul.go]: Remove line above and run goimports one more time

Comment (wendigo) [inline consul/consul.go]: Sure. Done.

File: consul/consul.go [hunk 11]
```diff
@@ -203,26 +204,53 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
-		Check:   c.marathonToConsulCheck(task, app.HealthChecks, serviceAddress),
+		Checks:  c.marathonToConsulChecks(task, app.HealthChecks, serviceAddress),
 	}, nil
 }
 
-func (c *Consul) marathonToConsulCheck(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) *consulapi.AgentServiceCheck {
-	//	TODO: Handle all types of checks
+func (c *Consul) marathonToConsulChecks(task *apps.Task, healthChecks []apps.HealthCheck, serviceAddress string) consulapi.AgentServiceChecks {
+	var checks consulapi.AgentServiceChecks = make(consulapi.AgentServiceChecks, 0, len(healthChecks))
+
```
Comment (janisz) [inline consul/consul.go]: Remove this empty line

File: consul/consul.go [hunk 12]
```diff
@@ -203,26 +204,53 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
-		if check.Protocol == "HTTP" {
-			return &consulapi.AgentServiceCheck{
-				HTTP: (&url.URL{
-					Scheme: "http",
-					Host:   fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex]),
-					Path:   check.Path,
-				}).String(),
+		switch check.Protocol {
+		case "HTTP", "HTTPS":
+
```
Comment (janisz) [inline consul/consul.go]: Remove this empty line

Comment (wendigo) [inline consul/consul.go]: Sure

File: consul/consul.go [hunk 13]
```diff
@@ -203,26 +204,53 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+			if parsedUrl, err := url.ParseRequestURI(check.Path); err == nil {
+				parsedUrl.Scheme = strings.ToLower(check.Protocol)
+				parsedUrl.Host = fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex])
+
```
Comment (janisz) [inline consul/consul.go]: Remove this empty line

File: consul/consul.go [hunk 14]
```diff
@@ -203,26 +204,53 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+					Warn(fmt.Sprintf("Could not parse provided path: %s", check.Path))
+			}
+		case "TCP":
+			checks = append(checks, &consulapi.AgentServiceCheck{
+				TCP:      fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex]),
 				Interval: fmt.Sprintf("%ds", check.IntervalSeconds),
 				Timeout:  fmt.Sprintf("%ds", check.TimeoutSeconds),
-			}
+			})
+
```
Comment (janisz) [inline consul/consul.go]: Remove this empty line

File: consul/consul.go [hunk 15]
```diff
@@ -203,26 +204,53 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
+			checks = append(checks, &consulapi.AgentServiceCheck{
+				Script:   check.Command.Value,
+				Interval: fmt.Sprintf("%ds", check.IntervalSeconds),
+				Timeout:  fmt.Sprintf("%ds", check.TimeoutSeconds),
+			})
+		default:
+			log.WithField("Id", task.AppID.String()).WithField("Address", serviceAddress).
+				Warn(fmt.Sprintf("Unrecognized check protocol %s", check.Protocol))
 		}
+
```
Comment (janisz) [inline consul/consul.go]: Remove this empty line

File: consul/consul.go (hunk 14 above)
Comment (wendigo) [inline consul/consul.go]: Oh cmon :)

File: consul/consul.go (hunk 13 above)
Comment (wendigo) [inline consul/consul.go]: This makes this block totally unreadable. I will leave it as it is.

File: consul/consul.go (hunk 11 above)
Comment (wendigo) [inline consul/consul.go]: It will make code unreadable.

File: consul/consul_test.go (hunk 3 above)
Comment (wendigo) [inline consul/consul_test.go]: Added test case with invalid path

File: consul/consul.go (hunk 11 above)
Comment (janisz) [inline consul/consul.go]: Ok, got it

File: consul/consul.go [hunk 16]
```diff
@@ -203,26 +204,49 @@ func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*c
...
 	for _, check := range healthChecks {
-		if check.Protocol == "HTTP" {
-			return &consulapi.AgentServiceCheck{
-				HTTP: (&url.URL{
-					Scheme: "http",
-					Host:   fmt.Sprintf("%s:%d", serviceAddress, task.Ports[check.PortIndex]),
-					Path:   check.Path,
-				}).String(),
+		switch check.Protocol {
+		case "HTTP", "HTTPS":
```
Comment (janisz) [inline consul/consul.go]: Is https healthcheck available in Marathon?

Comment (wendigo) [inline consul/consul.go]: Yes it is according to comment in https://github.com/mesosphere/marathon/issues/2758

**Comment by janisz** [comment]
:+1:

**Comment by janisz** [comment]
@wendigo One more thing. Could you squash your commits into one?

**Comment by wendigo** [comment]
Squashed.

**Comment by dankraw** [comment]
nice! :+1:

**Comment by janisz** [comment]
Thanks @wendigo !

**Comment by wendigo** [comment]
Yeah! :sunglasses:

**Pull request merged**